crate_universe/extension.bzl

```diff
@@ -160,16 +167,52 @@ def _crate_impl(module_ctx):
     cargo_bazel = get_cargo_bazel_runner(module_ctx)
     all_repos = []
     for mod in module_ctx.modules:
+        module_annotations = {}
+        repo_specific_annotations = {}
+        for annotation_tag in mod.tags.annotation:
+            annotation_dict = structs.to_dict(annotation_tag)
+            repositories = annotation_dict.pop("repositories")
+            crate = annotation_dict.pop("crate")
+
+            # The crate.annotation function can take in either a list or a bool.
+            # For the tag-based method, because it has type safety, we have to
+            # split it into two parameters.
+            if annotation_dict.pop("gen_all_binaries"):
+                annotation_dict["gen_binaries"] = True
+            annotation = _crate_universe_crate.annotation(**{
+                k: v
+                for k, v in annotation_dict.items()
+                # Tag classes can't take in None, but the function requires None
+                # instead of the empty values in many cases.
+                if v != "" and v != [] and v != {}
```

Comment:
This doesn't cover the boolean arguments, which feels problematic for the booleans which we allow setting on the `from_cargo` level and overriding per crate with annotations (e.g. `gen_build_script`).

Is there an issue upstream about not being able to use `None` in `tag_class`es?

In the mean time, I guess we should probably ban booleans in our `tag_class`es and follow rules_go's style of having the `tag_class` take a string here which may be one of "on", "off", or "auto" (default "auto"), and converting "auto" to None here?

Reply:
I couldn't find an issue so I filed https://github.com/bazelbuild/bazel/issues/20744 - could you add a reference to it in this comment?

Reply:
This is only an issue for `gen_build_script`. There are only two boolean parameters to annotation - `disable_pipelining` and `gen_build_script`, and `disable_pipelining` defaults to `False` in the annotation rather than None.